File: apps/macos/Sources/Taskly/Repositories/Repositories.swift
import Foundation

/// Smart-view types. `List` renders one user list (`listId`).
public enum TaskViewType: Sendable, Hashable {
    case all
    case today
    case planned
    case completed
    case list(Int)
}

/// Repository above SQLiteDatabase: validation + view dispatch + grouping.
public struct TaskRepository: Sendable {
    let db: SQLiteDatabase
    let i18n: I18nService

    public init(db: SQLiteDatabase, i18n: I18nService) {
        self.db = db
        self.i18n = i18n
    }

    /// View dispatch. Note: `all` with showCompleted=false returns only
    /// incomplete tasks (reference behavior).
    public func getTasksByView(
        _ viewType: TaskViewType, listId: Int? = nil, limit: Int = 1000, offset: Int = 0,
        showCompleted: Bool = false
    ) throws -> [TaskItem] {
        switch viewType {
        case .today:
            return showCompleted ? try db.getTodayTasksIncludingCompleted(limit: limit, offset: offset)
                : try db.getTodayTasks(limit: limit, offset: offset)
        case .planned:
            return showCompleted ? try db.getPlannedTasksIncludingCompleted(limit: limit, offset: offset)
                : try db.getPlannedTasks(limit: limit, offset: offset)
        case .all:
            return showCompleted ? try db.getAllTasksIncludingCompleted(limit: limit, offset: offset)
                : try db.getIncompleteTasks(limit: limit, offset: offset)
        case .completed:
            return try db.getCompletedTasks(limit: limit, offset: offset)
        case .list(let id):
            return showCompleted
                ? try db.getTasksByListIncludingCompleted(id, limit: limit, offset: offset)
                : try db.getTasksByList(id, limit: limit, offset: offset)
        }
    }

    @discardableResult
    public func addTask(_ task: TaskItem) throws -> Int {
        if let error = ValidationHelper.validateTaskText(task.text, i18n) {
            throw error
        }
        return try db.addTask(task)
    }

    @discardableResult
    public func updateTask(_ task: TaskItem) throws -> Int {
        if let error = ValidationHelper.validateTaskText(task.text, i18n) {
            throw error
        }
        return try db.updateTask(task)
    }

    @discardableResult
    public func toggleTaskCompleted(_ id: Int) throws -> Int {
        try db.toggleTaskCompleted(id)
    }

    @discardableResult
    public func setTaskCompleted(_ id: Int, _ completed: Bool) throws -> Int {
        try db.setTaskCompleted(id, completed)
    }

    @discardableResult
    public func deleteTask(_ id: Int) throws -> Int {
        try db.deleteTask(id)
    }

    public func searchTasks(_ keyword: String) throws -> [TaskItem] {
        if keyword.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return []
        }
        return try db.searchTasks(keyword.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    public func getTaskById(_ id: Int) throws -> TaskItem? {
        try db.getTaskById(id)
    }

    /// Group tasks by listId preserving encounter order.
    public func groupTasksByList(_ tasks: [TaskItem]) -> [(listId: Int, tasks: [TaskItem])] {
        var order: [Int] = []
        var groups: [Int: [TaskItem]] = [:]
        for task in tasks {
            if groups[task.listId] == nil {
                order.append(task.listId)
            }
            groups[task.listId, default: []].append(task)
        }
        return order.map { ($0, groups[$0]!) }
    }

    public func getTaskCountByList(_ listId: Int) throws -> Int { try db.getTaskCountByList(listId) }
    public func getIncompleteTaskCount() throws -> Int { try db.getIncompleteTaskCount() }
    public func getCompletedTaskCount() throws -> Int { try db.getCompletedTaskCount() }
    public func getTodayTaskCount() throws -> Int { try db.getTodayTaskCount() }
    public func getPlannedTaskCount() throws -> Int { try db.getPlannedTaskCount() }
}

/// List repository: validation + business helpers above SQLiteDatabase.
public struct ListRepository: Sendable {
    let db: SQLiteDatabase
    let i18n: I18nService

    public init(db: SQLiteDatabase, i18n: I18nService) {
        self.db = db
        self.i18n = i18n
    }

    public func getAllLists() throws -> [TodoList] { try db.getAllLists() }
    public func getListById(_ id: Int) throws -> TodoList? { try db.getListById(id) }
    public func getListByName(_ name: String) throws -> TodoList? { try db.getListByName(name) }

    @discardableResult
    public func addList(_ name: String, icon: String? = nil, color: Int? = nil) throws -> Int {
        if let error = ValidationHelper.validateListName(name, i18n) {
            throw error
        }
        return try db.addList(name: name.trimmingCharacters(in: .whitespaces), icon: icon, color: color)
    }

    @discardableResult
    public func updateList(_ id: Int, _ name: String, icon: String? = nil, color: Int? = nil,
                           clearIcon: Bool = false, clearColor: Bool = false) throws -> Int {
        if let error = ValidationHelper.validateListName(name, i18n) {
            throw error
        }
        return try db.updateList(
            id: id, name: name.trimmingCharacters(in: .whitespaces),
            icon: icon, color: color, clearIcon: clearIcon, clearColor: clearColor)
    }

    @discardableResult
    public func updateListIcon(_ id: Int, _ icon: String) throws -> Int {
        guard let list = try getListById(id) else {
            throw AppError("List not found", .validation)
        }
        return try db.updateList(id: id, name: list.name, icon: icon, color: list.color)
    }

    @discardableResult
    public func updateListColor(_ id: Int, _ color: Int) throws -> Int {
        guard let list = try getListById(id) else {
            throw AppError("List not found", .validation)
        }
        return try db.updateList(id: id, name: list.name, icon: list.icon, color: color)
    }

    @discardableResult
    public func deleteList(_ id: Int) throws -> Int { try db.deleteList(id: id) }

    /// Default list = the first one.
    public func getDefaultList() throws -> TodoList? {
        try getAllLists().first
    }
}

File: apps/windows/Taskly/Repositories/Repositories.cs
using Taskly.Data;
using Taskly.Models;
using Taskly.Services;

namespace Taskly.Repositories;

/// <summary>
/// Repositories above SQLiteDatabase: validation + view dispatch.
/// (ListPalette moved here from the legacy ListRepository file.)
/// </summary>
public sealed class TaskRepository
{
    private readonly SQLiteDatabase _db;
    private readonly I18nService _i18n;

    public TaskRepository(SQLiteDatabase db, I18nService i18n)
    {
        _db = db;
        _i18n = i18n;
    }

    /// <summary>View dispatch; 'all' with showCompleted=false = incomplete only.</summary>
    public async Task<List<TaskItem>> GetTasksByViewAsync(
        TaskViewType viewType, int? listId = null,
        int limit = 1000, int offset = 0, bool showCompleted = false)
    {
        return viewType switch
        {
            TaskViewType.Today => showCompleted
                ? await _db.GetTodayTasksIncludingCompletedAsync(limit, offset)
                : await _db.GetTodayTasksAsync(limit, offset),
            TaskViewType.Planned => showCompleted
                ? await _db.GetPlannedTasksIncludingCompletedAsync(limit, offset)
                : await _db.GetPlannedTasksAsync(limit, offset),
            TaskViewType.All => showCompleted
                ? await _db.GetAllTasksIncludingCompletedAsync(limit, offset)
                : await _db.GetIncompleteTasksAsync(limit, offset),
            TaskViewType.Completed => await _db.GetCompletedTasksAsync(limit, offset),
            TaskViewType.List when listId is not null => showCompleted
                ? await _db.GetTasksByListIncludingCompletedAsync(listId.Value, limit, offset)
                : await _db.GetTasksByListAsync(listId.Value, limit, offset),
            _ => await _db.GetIncompleteTasksAsync(limit, offset),
        };
    }

    public async Task<int> AddTaskAsync(TaskItem task)
    {
        var error = ValidationHelper.ValidateTaskText(task.Text, _i18n);
        if (error is not null)
        {
            throw new ArgumentException(error.Message);
        }

        return await _db.AddTaskAsync(task);
    }

    public async Task<int> UpdateTaskAsync(TaskItem task)
    {
        var error = ValidationHelper.ValidateTaskText(task.Text, _i18n);
        if (error is not null)
        {
            throw new ArgumentException(error.Message);
        }

        return await _db.UpdateTaskAsync(task);
    }

    public Task<int> ToggleTaskCompletedAsync(int id) => _db.ToggleTaskCompletedAsync(id);
    public Task<int> SetTaskCompletedAsync(int id, bool completed) => _db.SetTaskCompletedAsync(id, completed);
    public Task<int> DeleteTaskAsync(int id) => _db.DeleteTaskAsync(id);

    public async Task<List<TaskItem>> SearchTasksAsync(string keyword)
    {
        if (string.IsNullOrWhiteSpace(keyword))
        {
            return new List<TaskItem>();
        }

        return await _db.SearchTasksAsync(keyword.Trim());
    }

    /// <summary>Calendar view data (PRODUCT-SPEC §4b); the view groups and
    /// renders client-side.</summary>
    public Task<List<TaskItem>> GetTasksInRangeAsync(string startDate, string endDate, bool includeCompleted = false) =>
        _db.GetTasksInRangeAsync(startDate, endDate, includeCompleted);

    public Task<List<DueDayCount>> GetDueDayCountsAsync(string startDate, string endDate) =>
        _db.GetDueDayCountsAsync(startDate, endDate);

    public Task<TaskItem?> GetTaskByIdAsync(int id) => _db.GetTaskByIdAsync(id);

    public Task<int> GetTaskCountByListAsync(int listId) => _db.GetTaskCountByListAsync(listId);
    public Task<int> GetIncompleteTaskCountAsync() => _db.GetIncompleteTaskCountAsync();
    public Task<int> GetCompletedTaskCountAsync() => _db.GetCompletedTaskCountAsync();
    public Task<int> GetTodayTaskCountAsync() => _db.GetTodayTaskCountAsync();
    public Task<int> GetPlannedTaskCountAsync() => _db.GetPlannedTaskCountAsync();
}

public sealed class ListRepository
{
    private readonly SQLiteDatabase _db;
    private readonly I18nService _i18n;

    public ListRepository(SQLiteDatabase db, I18nService i18n)
    {
        _db = db;
        _i18n = i18n;
    }

    public Task<List<TodoList>> GetAllListsAsync() => _db.GetAllListsAsync();
    public Task<TodoList?> GetListByIdAsync(int id) => _db.GetListByIdAsync(id);
    public Task<TodoList?> GetListByNameAsync(string name) => _db.GetListByNameAsync(name);

    public async Task<int> AddListAsync(string name, string? icon = null, int? color = null)
    {
        var error = ValidationHelper.ValidateListName(name, _i18n);
        if (error is not null)
        {
            throw new ArgumentException(error.Message);
        }

        return await _db.AddListAsync(name.Trim(), icon, color);
    }

    public async Task<int> UpdateListAsync(int id, string name, string? icon = null, int? color = null,
        bool clearIcon = false, bool clearColor = false)
    {
        var error = ValidationHelper.ValidateListName(name, _i18n);
        if (error is not null)
        {
            throw new ArgumentException(error.Message);
        }

        return await _db.UpdateListAsync(id, name.Trim(), icon, color, clearIcon, clearColor);
    }

    public async Task<int> DeleteListAsync(int id) => await _db.DeleteListAsync(id);

    /// <summary>Default list = the first one.</summary>
    public async Task<TodoList?> GetDefaultListAsync()
    {
        var lists = await GetAllListsAsync();
        return lists.Count > 0 ? lists[0] : null;
    }
}

/// <summary>Preset palette + emoji categories (fixed order, contract).</summary>
public static class ListPalette
{
    public static readonly string[] Colors =
    {
        "#007AFF", "#FF3B30", "#FF9500", "#FFCC00", "#4CD964",
        "#5AC8FA", "#5856D6", "#FF2D55", "#8E8E93", "#C7C7CC",
    };

    public static readonly string[][] EmojiCategories =
    {
        new[] { "📋", "📝", "✅", "🎯", "💡", "📌", "🔖", "📎" },
        new[] { "🏠", "🏢", "💼", "📱", "💻", "🎨", "📚", "🎓" },
        new[] { "❤️", "⭐", "🌟", "🔥", "💪", "🎉", "🎊", "🏆" },
        new[] { "🛒", "🛍️", "🍔", "☕", "🍕", "🥤", "🎮", "🎬" },
        new[] { "✈️", "🚗", "🚴", "🏃", "⚽", "🏀", "🎸", "🎵" },
        new[] { "💰", "💳", "📊", "📈", "💼", "📧", "📅", "⏰" },
    };
}
